File: 1408-find-the-smallest-divisor-given-a-threshold/1408-find-the-smallest-divisor-given-a-threshold.java
class Solution {
    public int smallestDivisor(int[] nums, int threshold) {
        int right = Arrays.stream(nums).max().getAsInt();
        int left = 1;
        int ctr = 0;
        while (left < right){
            int mid = left + (right - left)/2;
            System.out.print("counter :  "+ ctr++);
            System.out.println(mid);
            if(check(mid, threshold, nums)){
                right = mid;
            }else{
                left = mid + 1;
            }
        }
        return left;
    }
    private boolean check(int mid, int threshold, int[] nums){
        int ans = 0;
        for(int num : nums){
            ans += (num + mid - 1)/ mid;
        }
        return ans<= threshold;
    }
}

File: 1408-find-the-smallest-divisor-given-a-threshold/1408-find-the-smallest-divisor-given-a-threshold.rs
impl Solution {
    pub fn smallest_divisor(nums: Vec<i32>, threshold: i32) -> i32 {
        let &max_val = nums.iter().max().unwrap();
        let (mut left, mut right) = (1, max_val);
        while left < right {
            let mid = left + (right - left) / 2;
            let sum: i32 = nums.iter().map(|&num| (num + mid - 1) / mid).sum();
            if sum <= threshold {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        left
    }
}
